### PR TITLE
[Core][Spilled Object Leakage] More robust spilled object deletion

### LINKED ISSUE #26395: [Core] raylet crashed due to unhandled std::length_error
### What happened + What you expected to happen

For chaos_dataset_shuffle_sort_1tb, raylet crashes due to unhandled std::length_error. The cause is that we are trying to create a string that exceeds the maximum size.
The code is
```
bool SpilledObjectReader::ParseObjectHeader(std::istream &is,
                                            uint64_t object_offset,
                                            uint64_t &data_offset,
                                            uint64_t &data_size,
                                            uint64_t &metadata_offset,
                                            uint64_t &metadata_size,
                                            rpc::Address &owner_address) {
  if (!is.seekg(object_offset)) {
    return false;
  }

  uint64_t address_size = 0;
  if (!ReadUINT64(is, address_size) || !ReadUINT64(is, metadata_size) ||
      !ReadUINT64(is, data_size)) {
    return false;
  }

  std::string address_str(address_size, '\0'); <----------- address_size is 4699037094196886216
```

Seems the spilled file is corrupted.

### Versions / Dependencies

master

### Reproduction script

chaos_dataset_shuffle_sort_1tb

### Issue Severity

_No response_

### LINKED ISSUE #26994: [Core] spilled object leaked.
### What happened + What you expected to happen

Run the workload specified in https://gist.github.com/scv119/5339b599ed14d010437c5a12d7aed96a with larger dataset (i.e. 200GB data) and cancel the job while it's spilling. you will noticed that the spilled file is leaked, even `ray memory` reports no objects.

### Versions / Dependencies

latest ray

### Reproduction script

N/A

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by scv119**
## Why are these changes needed?

We have noticed spilled objects not deleted even if the job creating those objects finished execution. After reading the code my theory is that the object delegation worker failed in the middle of deleting spilled files, which doesn't handle well in today's spilled object deletion logic.

Though I no longer get a reproduction, (which I suspect due to the fix https://github.com/ray-project/ray/issues/26395), we can enhance the failure handle logic when object deletion failed.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scv119** [comment]
TBD, add tests.

**Comment by rkooo567** [comment]
```
my theory is that the object delegation worker failed in the middle of deleting spilled files, which doesn't handle well in today's spilled object deletion logic.
```

Is it something we can reproducing from cpp tests or by killing the spill worker in the middle of the test?

File: src/ray/raylet/local_object_manager.h [hunk 1]
```diff
@@ -213,12 +219,18 @@ class LocalObjectManager {
   /// Delete spilled objects stored in given urls.
   ///
   /// \param urls_to_delete List of urls to delete from external storages.
-  void DeleteSpilledObjects(std::vector<std::string> &urls_to_delete);
+  /// \param num_retries Num of retries allowed in case of failure, zero or negative
+  /// means don't retry.
+  void DeleteSpilledObjects(std::vector<std::string> urls_to_delete,
+                            int64_t num_retries = kDefaultSpilledObjectDeleteRetries);
```
Comment (rkooo567) [inline src/ray/raylet/local_object_manager.h:225]: num_retries_left?

File: src/ray/raylet/local_object_manager.h [hunk 2]
```diff
@@ -365,6 +377,9 @@ class LocalObjectManager {
   /// The last time a restore log finished.
   int64_t last_restore_log_ns_ = 0;
 
+  /// The number of failed deletion requests.
+  std::atomic<int64_t> num_failed_deletion_requests_ = 0;
```
Comment (rkooo567) [inline src/ray/raylet/local_object_manager.h:381]: Btw I think this doesn't have to be atomic (it is single threaded)

File: src/ray/raylet/local_object_manager.cc [hunk 3]
```diff
@@ -613,6 +624,8 @@ void LocalObjectManager::RecordMetrics() const {
   ray::stats::STATS_spill_manager_request_total.Record(spilled_objects_total_, "Spilled");
   ray::stats::STATS_spill_manager_request_total.Record(restored_objects_total_,
                                                        "Restored");
+  ray::stats::STATS_spill_manager_request_total.Record(num_failed_deletion_requests_,
```
Comment (rkooo567) [inline src/ray/raylet/local_object_manager.cc:627]: Hmm Wonder if we need more breakdown;;

type: spill/restore/delete
status: success/failed

File: python/ray/_private/external_storage.py [hunk 4]
```diff
@@ -331,7 +331,11 @@ def restore_spilled_objects(
     def delete_spilled_objects(self, urls: List[str]):
         for url in urls:
             path = parse_url_with_offset(url.decode()).base_url
-            os.remove(path)
+            try:
```
Comment (rkooo567) [inline python/ray/_private/external_storage.py:334]: Can you update the ABC method definition this method should handle worker failure & retry case?

**Review by rkooo567 — APPROVED** [review]
Mostly LGTM!

File: src/ray/raylet/local_object_manager.h (hunk 2 above)
Comment (scv119) [reply src/ray/raylet/local_object_manager.h:381]: hmm i thought the ` io_worker->rpc_client()->DeleteSpilledObjects` callback happens in a different thread.

File: src/ray/raylet/local_object_manager.cc (hunk 3 above)
Comment (rickyyx) [reply src/ray/raylet/local_object_manager.cc:640]: i could work on this as part of 2.2.

File: src/ray/raylet/local_object_manager.h (hunk 2 above)
Comment (rickyyx) [reply src/ray/raylet/local_object_manager.h:388]: maybe we should start document explicitly where callbacks are run w.r.t threading model lol

**Review by rickyyx — APPROVED** [review]
Are there any release tests we need to trigger for this? Or it was originally from some other failures.

**Pull request merged**